### DISCUSSION TITLE
Artifactory returning 404 for npm modules that exist.

### DISCUSSION THREAD

**Discussion by dylang** [renovatebot/renovate#21703] [Suggest an Idea]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

31.97.0

### Please select which platform you are using if self-hosting.

GitHub Enterprise Server

### If you're self-hosting Renovate, tell us what version of the platform you run.

3.3.1

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the bug

Hi! 

* All renovate npm module PR's that haven't been modified were autoclosed.
* All dashboards are empty except for lockfile updates.
* Getting 404's in logs when attempting to look up npm packages.

I think the issue might be caused by this PR:

https://github.com/renovatebot/renovate/pull/14424/files

And maybe this is the problem:

Before this PR:
```ts
const body = (await http.getJson<NpmResponse>(packageUrl, { headers })).body;
```

After this PR:
```ts
const body = (await http.getJson<NpmResponse>(packageUrl)).body;
```

Is it possible we're no longer sending the auth headers? These are required for our Artifactory server.

I haven't made a ticket for a while because everything has been so smooth for years, or you fixed issues before I could report them! Thanks for the amazing work.

### Relevant debug logs

<details><summary>Logs</summary>

```
{"err":{"name":"HTTPError","code":"ERR_NON_2XX_3XX_RESPONSE","timings":{"start":1646192901209,"socket":1646192901209,"lookup":1646192901210,"connect":1646192901255,"secureConnect":1646192901347,"upload":1646192901347,"response":1646192901398,"end":1646192901399,"phases":{"wait":0,"dns":1,"tcp":45,"tls":92,"request":0,"firstByte":51,"download":1,"total":190}},"message":"Response code 404 (Not Found)","stack":"HTTPError: Response code 404 (Not Found)\n    at Request.<anonymous> (/home/jenkins/workspace/lia_company-org_renovate_master/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)","options":{"headers":{"user-agent":"RenovateBot/31.94.0 (https://github.com/renovatebot/renovate)","accept":"application/json","accept-encoding":"gzip, deflate, br"},"url":"https://artifactory.eng.company.com/api/npm/virtual-npm/renovate","hostType":"npm","username":"","password":"","method":"GET","http2":false},"response":{"statusCode":404,"statusMessage":"Not Found","body":{"error":"not_found","reason":"document not found"},"headers":{"date":"Wed, 02 Mar 2022 03:48:21 GMT","content-type":"application/json","transfer-encoding":"chunked","connection":"close","server":"Artifactory/6.18.1","x-artifactory-id":"c6c1d1bfa66d40223d7aadd71901d53578a354f2","x-artifactory-node-id":"art-ip-10-251-0-188","content-encoding":"gzip"},"httpVersion":"1.1"}},"packageName":"renovate","msg":"Dependency lookup failure: not found","time":"2022-03-02T03:48:21.401Z","v":0}
{"err":{"name":"HTTPError","code":"ERR_NON_2XX_3XX_RESPONSE","timings":{"start":1646192901208,"socket":1646192901208,"lookup":1646192901210,"connect":1646192901254,"secureConnect":1646192901346,"upload":1646192901346,"response":1646192901400,"end":1646192901400,"phases":{"wait":0,"dns":2,"tcp":44,"tls":92,"request":0,"firstByte":54,"download":0,"total":192}},"message":"Response code 404 (Not Found)","stack":"HTTPError: Response code 404 (Not Found)\n    at Request.<anonymous> (/home/jenkins/workspace/lia_company-org_renovate_master/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)","options":{"headers":{"user-agent":"RenovateBot/31.94.0 (https://github.com/renovatebot/renovate)","accept":"application/json","accept-encoding":"gzip, deflate, br"},"url":"https://artifactory.eng.company.com/api/npm/virtual-npm/lerna","hostType":"npm","username":"","password":"","method":"GET","http2":false},"response":{"statusCode":404,"statusMessage":"Not Found","body":{"error":"not_found","reason":"document not found"},"headers":{"date":"Wed, 02 Mar 2022 03:48:21 GMT","content-type":"application/json","transfer-encoding":"chunked","connection":"close","server":"Artifactory/6.18.1","x-artifactory-id":"d7d4e5a1b8ab6e8c0ea8899a2e98ded8158429b2","x-artifactory-node-id":"art-ip-10-251-0-157","content-encoding":"gzip"},"httpVersion":"1.1"}},"packageName":"lerna","msg":"Dependency lookup failure: not found","time":"2022-03-02T03:48:21.402Z","v":0}

```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by rarkins** [comment]
hostRules should be adding those headers later in the process. Can you describe how/where you configure the npmrc files in your repo or config?

**Comment by dylang** [comment]
This configuration hasn't changed in months. 

```
_auth =<BASE64 USERNAME:APIKEY>==
email = email@company.com
registry=https://artifactory.company.com/api/npm/virtual-npm
always-auth=true

@prefix:registry=https://artifactory.company.com/api/npm/virtual-npm/
//artifactory.company.com/api/npm/virtual-npm/:_password=<BASE64 APIKEY>
//artifactory.company.com/api/npm/virtual-npm/:username=username
//artifactory.company.com/api/npm/virtual-npm/:email=username@company.com
//artifactory.company.com/api/npm/virtual-npm/:always-auth=true
## We also have our deprecated artifactory hostname, but it's not needed for the issues we're seeing.
//company.io/api/npm/virtual-npm/:_password=<BASE64 APIKEY>
//company.io/api/npm/virtual-npm/:username=username
//company.io/api/npm/virtual-npm/:email=username@company.com
//company.io/api/npm/virtual-npm/:always-auth=true
```

Maybe related? When artifactory is hit with too many simultaneous requests it seems to 404 error codes. When this happens, if I try again later it will work, but I've had the above error happen every hour for the last few days.

**Comment by rarkins** [comment]
Where/how is that configured?

**Comment by rarkins** [comment]
Also is it intentional that you have some duplication of both registry and your auth within that file? Do you know if it's somehow *necessary*?

**Comment by dylang** [comment]
That's from the `.npmrc` on the machine - I don't think the team that set this up are users of node/npm/renovate, and I don't know why it's lik that. I've tried finding a definitive guide to what the `.npmrc` should look like for multiple private Artifactory repositories and haven't found one. Is there a guide you recommend?

I just ran `yarn renovate` locally with all of the same settings and it worked fine (no 404), so I'm thinking the problem is with our Artifactory. I'm fine closing this issue, thanks for the help!

**Comment by rarkins** [comment]
I'd like to try to understand it better before closing.

Are you sure you don't have any hostRules or npmrc settings in Renovate config?

**Comment by dylang** [comment]
Oh, we do:

```
    npmrc: readFileSync(path.join(homedir(), '.npmrc'), 'utf8'),

    packageRules: [
        {
            matchDatasources: ['docker'],
            registryUrls: ['https://virtual-docker.artifactory.company.com'],
        }
    ],

    hostRules: [
        {
            hostType: 'docker',
            matchHost: 'virtual-docker.artifactory.company.com',
            username: process.env.ARTIFACTORY_USERNAME,
            password: process.env.ARTIFACTORY_PASSWORD,
        },
        {
            hostType: 'docker',
            matchHost: 'virtual-docker.company.io',
            username: process.env.ARTIFACTORY_USERNAME,
            password: process.env.ARTIFACTORY_PASSWORD,
        }
    ],
```

I have some feature requests from this that I can create as separate tickets if you think they would be helpful.

* The http failures don't show up in the default console output. I see them in the verbose logs only. I think they are log level 20, but I forget what that is.
* If we count this as a failure:  https://github.com/renovatebot/renovate/issues/7376
*  If possible, don't remove PR's because of these errors, just ignore them. That way a flaky connection or server doesn't auto-close all of the PRs. :)
* When autoclosing, add a comment explaining why it is being closed. Ideally, include the error message if there is one.
* Show most recent failures on the dashboard. This should help us know when an api key is no longer working, or we're being rate limited.

**Comment by rarkins** [comment]
> Oh, we do:
> 
> ```
>     npmrc: readFileSync(path.join(homedir(), '.npmrc'), 'utf8'),
> 
>     packageRules: [
>         {
>             matchDatasources: ['docker'],
>             registryUrls: ['https://virtual-docker.artifactory.company.com'],
>         }
>     ],
> 
>     hostRules: [
>         {
>             hostType: 'docker',
>             matchHost: 'virtual-docker.artifactory.company.com',
>             username: process.env.ARTIFACTORY_USERNAME,
>             password: process.env.ARTIFACTORY_PASSWORD,
>         },
>         {
>             hostType: 'docker',
>             matchHost: 'virtual-docker.company.io',
>             username: process.env.ARTIFACTORY_USERNAME,
>             password: process.env.ARTIFACTORY_PASSWORD,
>         }
>     ],
> ```

Thanks. So no `hostRules` with `hostType` blank or `hostType=npm`?

> I have some feature requests from this that I can create as separate tickets if you think they would be helpful.
> 
> * The http failures don't show up in the default console output. I see them in the verbose logs only. I think they are log level 20, but I forget what that is.

Sounds like debug level. Currently that's intentional, but happy to discuss it. Essentially our problem is that a lot of adopters of Renovate "didn't mind" it failing for private repos, and are happy for it to be suppressed. We need to reevaluate though.

> * If we count this as a failure:  [Fatal log messages have debug log level #7376](https://github.com/renovatebot/renovate/issues/7376)

Kind of same as above. One person's "this should be fatal" is another person's "just update what you can, bot"

> * If possible, don't remove PR's because of these errors, just ignore them. That way a flaky connection or server doesn't auto-close all of the PRs. :)

You can enable that option already: https://docs.renovatebot.com/configuration-options/#prunestalebranches

> * When autoclosing, add a comment explaining why it is being closed. Ideally, include the error message if there is one.

I know it makes sense as a requirement, but the reality is that Renovate doesn't really "know" why a branch no longer exists.

> * Show most recent failures on the dashboard. This should help us know when an api key is no longer working, or we're being rate limited.

I think it could be good to list how many dependencies are failing to look up. This can definitely be a feature request for the dashboard

**Comment by rarkins** [comment]
Here's how I'm testing:

My `config.npmrc`:

```
_auth = dXNlcjE6a2V5MQ==
email = email@company.com
registry=https://artifactory.company.com/api/npm/virtual-npm
always-auth=true

@prefix:registry=https://artifactory.company.com/api/npm/virtual-npm/
//artifactory.company.com/api/npm/virtual-npm/:_password=a2V5Mg==
//artifactory.company.com/api/npm/virtual-npm/:username=user2
//artifactory.company.com/api/npm/virtual-npm/:email=username@company.com
//artifactory.company.com/api/npm/virtual-npm/:always-auth=true
```

`dXNlcjE6a2V5MQ==` is a base64 encoding of `user1:key1`.
`a2V5Mg==` is a base64 encoding of `key2`.

No `.npmrc` in the repo.

Now I roll back to `31.81.0` (which I assume is a working version for you) and debug a HTTP call:

![image](https://user-images.githubusercontent.com/6311784/156556277-539dea5b-a1ef-4c72-9a46-c80333571de0.png)

This shows:
- artifactory is being attempted
- Basic auth is applied
- The Basic auth value is the base64 encoding of `us3r2:passw0rd2`

This is expected, because I think the second section makes the initial `_auth=` line pointless.

Now I run with `31.97.2` (latest version), and I get the same:

![image](https://user-images.githubusercontent.com/6311784/156557012-8e0f1f38-ff09-4291-a59d-a8194c8ba58c.png)

@dylang this leaves two leading explanations I can think of:

1. Your Artifactory is causing the problem.. maybe it's an unlucky coincidence? I usually don't believe in them though
2. There's some other `npmrc`/`hostRules` interference, perhaps e.g. `.npmrc` in repos?

**Comment by jannylund** [comment]
We noticed similar behavior:

### Previous working setup
- Renovate running via docker, image `renovate/renovate:latest`
- NPM configured via the environment variable `RENOVATE_NPMRC`, completely overriding registry with our Artifactory and auth configuration.

### Problem 1
As of a some days ago (not sure on exact release), we suddenly noticed that all frontend PRs and branches disappeared.
- Enabling debug logs showed that renovate threw 401 on any contact to Artifactory for npm.

### Solution
- Migrating configuration from `RENOVATE_NPMRC` to `hostRules` fixed major part of the projects.

### Problem 2
Some projects started throwing artifact errors that it could not update lock file. Enabling debug logs shows that projects with multiple package.json files in subfolders fail on any subsequent update.


### Solution
- Restoring `RENOVATE_NPMRC` solves this.

So right now, we need to have double configurations. Auth config from `hostRules` are used for initial `npm install` / `yarn install` but `RENOVATE_NPMRC` is used for the subfolders. 🤯 

EDIT: Potentially related, npm handles npmrc differently if forked from within a node context https://github.com/npm/npm/issues/17674

**Comment by dylang** [comment]
I've made these changes and that particular problem is fixed, but maybe that wasn't the real issue causing the PR's to all close.

The new config:

```
    npmrc: `registry=https://artifactory.company.com/api/npm/virtual-npm\nalways-auth=true`,

    packageRules: [
        {
            matchDatasources: ['docker'],
            registryUrls: ['https://virtual-docker.company.com'],
        },
        {
            matchDatasources: ['gradle'],
            matchPackagePatterns: ['company'],
            registryUrls: ['https://artifactory.company.com/virtual-gradle'],
        },
    ],

    hostRules: [
        {
            hostType: 'npm',
            username: process.env.ARTIFACTORY_USERNAME,
            password: process.env.ARTIFACTORY_PASSWORD
        },
        {
            hostType: 'maven',
            username: process.env.ARTIFACTORY_USERNAME,
            password: process.env.ARTIFACTORY_PASSWORD
        },
        {
            hostType: 'gradle',
            matchHost: ['artifactory.company.com'],
            username: process.env.ARTIFACTORY_USERNAME,
            password: process.env.ARTIFACTORY_PASSWORD
        },
        {
            hostType: 'docker',
            matchHost: 'virtual-docker.company.com',
            username: process.env.ARTIFACTORY_USERNAME,
            password: process.env.ARTIFACTORY_PASSWORD
        },
        {
            hostType: 'docker',
            matchHost: 'virtual-docker.company.io',
            username: process.env.ARTIFACTORY_USERNAME,
            password: process.env.ARTIFACTORY_PASSWORD
        }
    ],
```

Node projects only use `yarn`, never `npm`, so that might be why I'm not seeing @jannylund's issue.

As it runs it's re-opening closed PR's.

But all is not perfect. I am seeing this error, I dont know if it's new or just exposed because the other failures were preventing it from getting this far or I just didn't know it was happening:

```json
        "response": {
           "statusCode": 422,
           "statusMessage": "Unprocessable Entity",
           "body": {
             "message": "Reference already exists",
             "documentation_url": "https://docs.github.com/enterprise/3.3/rest/reference/git#create-a-reference"
           },
```
immediately followed by:
```
DEBUG: Could not recreate autoclosed branch - skipping reopen
```

and later on...

```
DEBUG: gitRetry round 5 (repository=**REPO**, branch=renovate/selenium-standalone-7.x)
DEBUG: Git function thrown (repository=**REPO**, branch=renovate/selenium-standalone-7.x)
       "err": {
         "task": {
           "commands": [
             "push",
             "origin",
             "renovate/selenium-standalone-7.x:renovate/selenium-standalone-7.x",
             "--force-with-lease",
             "-u",
             "--no-verify",
             "--verbose",
             "--porcelain"
           ],
           "format": "utf-8",
           "parser": "[function]"
         },
         "message": "To https://github.company.com/org/project.git\n!\trefs/heads/renovate/selenium-standalone-7.x:refs/heads/renovate/selenium-standalone-7.x\t[rejected] (stale info)\nDone\nPushing to https://github.company.com/ci-org/argus.git\nerror: failed to push some refs to 'https://github.company.com/ci-org/argus.git'\n",
         "stack": "Error: To https://github.company.com/ci-org/argus.git\n!\trefs/heads/renovate/selenium-standalone-7.x:refs/heads/renovate/selenium-standalone-7.x\t[rejected] (stale info)\nDone\nPushing to https://github.company.com/org/project.git\nerror: failed to push some refs to 'https://github.company.com/org/project.git'\n\n    at Object.action (/Users/dylang/company/renovate/node_modules/simple-git/dist/cjs/index.js:1194:25)\n    at PluginStore.exec (/Users/dylang/company/renovate/node_modules/simple-git/dist/cjs/index.js:1229:29)\n    at /Users/dylang/company/renovate/node_modules/simple-git/dist/cjs/index.js:1589:43\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/Users/dylang/company/renovate/node_modules/simple-git/dist/cjs/index.js:1587:16)\n    at GitExecutorChain.<anonymous> (/Users/dylang/company/renovate/node_modules/simple-git/dist/cjs/index.js:1571:44)\n    at Generator.next (<anonymous>)\n    at fulfilled (/Users/dylang/company/renovate/node_modules/simple-git/dist/cjs/index.js:55:24)"
       }
```

Nobody is touching the renovate PRs except me, and I'm just running one copy of Renovate. 

After these errors seems to close the PR, but I don't see anything in the log saying the PRs are being closed.

GHE is configured with environment variables only:
```
GITHUB_ENDPOINT=https://github.company.com/api/v3
RENOVATE_ENDPOINT=https://github.company.com/api/v3
RENOVATE_TOKEN_USERNAME=....
RENOVATE_TOKEN=....
```

Maybe it's caused by the PRs being closed from the previous errors, and then re-opened?

This is happening to every PR is most repos, but some repos are not affected. None of the repos have `.npmrc` or `.yarnrc` in them.

**Comment by dylang** [comment]
I've been running all weekend with no problems. I added `repositoryCache: 'reset'` (it was previously not set) just in case it was the experimental cache causing the problem.

We run Renovate as a Jenkins task, and maybe it's possible that while `main` is running, a PR for an updated version of `renovate` is also running, causing corruption to the repo cache?

**Comment by rarkins** [comment]
Hopefully not. Also, repositoryCache does not store lookup data - only the extract results. BTW once you've reset it once, you should change it back to enabled

**Comment by JennToo** [comment]
I wanted to chime in that we were seeing this problem as well, in a self-hosted Renovate environment. The workaround of including a `hostRule` with the appropriate `token` and `authType` seemed to fix it. This is the relevant excerpt from our config:

```javascript
let config = {
  // This was here already, and I've kept it
  npmrc: secrets.npmrc,

  // Adding this fixed the issues we were seeing
  hostRules: [
    {
      hostType: 'npm',
      matchHost: 'artifactory.example.com',
      token: secrets.npmToken,
      authType: 'Bearer',
    },
  ],
  // ...
}
```

**Comment by viceice** [comment]
> I wanted to chime in that we were seeing this problem as well, in a self-hosted Renovate environment. The workaround of including a `hostRule` with the appropriate `token` and `authType` seemed to fix it. This is the relevant excerpt from our config:
> 
> ```javascript
> let config = {
>   // This was here already, and I've kept it
>   npmrc: secrets.npmrc,
> 
>   // Adding this fixed the issues we were seeing
>   hostRules: [
>     {
>       hostType: 'npm',
>       matchHost: 'artifactory.example.com',
>       token: secrets.npmToken,
>       authType: 'Bearer',
>     },
>   ],
>   // ...
> }
> ```

remove `authType`, as it's the default value.

**Comment by JennToo** [comment]
Strangely enough, I removed `authType` and our npm updates stopped working (we've got some automated internal tests that validate our renovate config with some dry-run scans). Not really sure why it's necessary, but putting it back fixed the issue. Version 32.35.0

**Comment by viceice** [comment]
> Strangely enough, I removed `authType` and our npm updates stopped working (we've got some automated internal tests that validate our renovate config with some dry-run scans). Not really sure why it's necessary, but putting it back fixed the issue. Version 32.35.0

That's strange. I know it works for me without that

default `Bearer`
https://github.com/renovatebot/renovate/blob/030b1a61a45a98557ddfe22732ffa1807619cba5/lib/util/http/auth.ts#L48-L55

**Comment by JennToo** [comment]
Interesting. I went ahead and put a breakpoint in there (with `authType` removed from the config) and it looks like something from a higher level has already configured the type to `Basic` so the `'Bearer'` type doesn't kick in here:

![image](https://user-images.githubusercontent.com/65889/168109365-3d2b0836-6082-4703-b895-5ede1c29c709.png)

I'll see if I can figure out where that `Basic` came from

**Comment by JennToo** [comment]
Aha! I tracked it down to this area:

https://github.com/renovatebot/renovate/blob/789cf38d8b416bd3f79fdabacd7aa2334fe23e49/lib/modules/datasource/npm/npmrc.ts#L68-L70

Turns out our npmrc file has `_auth` even though it's a token and should be using `_authToken`. That's why I had to add the overridden host rule. It's a pretty old file, so maybe it's using an old format or something.

If I change our npmrc file to use `_authToken` it does the right thing and I don't need an explicit host rule at all

**Comment by JennToo** [comment]
I sent that too early, our current npmrc file has both `_auth` and `_authToken` which might be why there was a problem

**Comment by argos83** [comment]
I'm experiencing a similar problem fetching NPM presets from artifactory (currently testing with 32.50.2) . It used to work before https://github.com/renovatebot/renovate/pull/14424/files

A few things that may help find the cause.
1)  Fetching NPM presets works for the the first repository in the list, the subsequent repositories fail with a 401 response. It's not an artifactory issue (like rate-limiting or smth), running separate parallel process with one repo each does work.
2) changing the npm host rules from username/password to token auth seems to fix the issue

So it may be that the state of the client is altered at some point after processing the first repo and the user/pass credentials are somewhat lost.
